### PR TITLE
[DOC-20] Document class file content for JBMC / java_bytecode readme

### PR DESCRIPTION
**Pull request by mgudemann**

### CODE REVIEW

**Review by peterschrammel — COMMENTED** [review]
What about debugging info?

File: jbmc/src/java_bytecode/README.md [hunk 1]
```diff
@@ -25,7 +25,80 @@ To be documented.
...
+files.
+
+There exists an [official
+specification](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html)
+
+Each class files contains information about its version, the constant pool,
+information about the contained class, its super-class, its implemented
+interfaces, its fields, methods and finally additional attributes.
+
+The content of a .class file can be inspected via the `javap` tool which is part
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: special mentiond for `javap -c`?

File: jbmc/src/java_bytecode/README.md [hunk 2]
```diff
@@ -25,7 +25,80 @@ To be documented.
 
 \subsection java-class-section How a java program / class is represented in a .class
 
-To be documented.
+Every Java class is compiled into a .class file, inner classes, anonymous
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: Clearer as for example "Every Java class is compiled into a .class file. Inner classes, anonymous classes or classes for tableswitches are also compiled into their own .class files."

File: jbmc/src/java_bytecode/README.md [hunk 3]
```diff
@@ -25,7 +25,80 @@ To be documented.
...
+classes or classes for tableswitches are compiled into their own .class
+files.
+
+There exists an [official
+specification](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html)
+
+Each class files contains information about its version, the constant pool,
+information about the contained class, its super-class, its implemented
+interfaces, its fields, methods and finally additional attributes.
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: maybe append "... such as exceptions thrown by a method or inner classes the class contains."

File: jbmc/src/java_bytecode/README.md [hunk 4]
```diff
@@ -25,7 +25,80 @@ To be documented.
...
+
+The content of a .class file can be inspected via the `javap` tool which is part
+of the JDK installation.
+
+In general, all variable length entries in a class file are represented by first
+having an integer `n` that specified the number of entries and then an array of
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: specifies

File: jbmc/src/java_bytecode/README.md [hunk 5]
```diff
@@ -25,7 +25,80 @@ To be documented.
...
+`n` such entries in the file.
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: Might also be worth noting that variable length attributes have an attribute length in bytes?

File: jbmc/src/java_bytecode/README.md [hunk 6]
```diff
@@ -25,7 +25,80 @@ To be documented.
...
+
+The integer format used in class files are unsigned integers of size 8/16/32
+bit, which are named as `u1`/`u2`/`u4`.
+
+\subsection Access flags
+
+The JVM specification defines different access flags, e.g., `final`, `static`,
+`protected`, `private` etc. where different ones are applicable to the class
+itself, its fields or methods. All access flags are represented as bits, the set
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/README.md]: This sentence is not terribly clear to me. Consider something like "All access flags are represented as constant u2 values whose flag names are prefixed with`ACC_`, for example, ACC_PUBLIC is 0x0001.

**Review by jeannielynnmoulton — APPROVED** [review]
Can be merged as is. I've made a few suggestions for where I think it could be made clearer.

**Comment by mgudemann** [comment]
@peterschrammel everything "optional" is more or less debugging information, will state this more explicitly

File: jbmc/src/java_bytecode/README.md (hunk 1 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/README.md]: I mentioned the options I use most of the time, `-c -p -l -verbose`

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**